File: bin2csv.py
import struct

def convert_binary_to_csv(binary_file_path, csv_file_path):
    with open(binary_file_path, 'rb') as bin_file:
        # Read the number of columns from the first line
        num_columns = int(bin_file.readline().decode().strip())

        # Read the column headers from the second line
        column_headers = bin_file.readline().decode().strip().split()

        # Check if the number of columns matches the number of headers
        if len(column_headers) != num_columns:
            raise ValueError("Number of columns does not match number of headers")

        with open(csv_file_path, 'w') as csv_file:
            # Write the column headers to the CSV file
            csv_file.write(','.join(column_headers) + '\n')

            # Read and write the binary data
            while True:
                # Read 32-bit floats from binary file
                data = bin_file.read(4 * num_columns)

                if not data:
                    break

                # Unpack binary data into floats
                values = struct.unpack(f'{num_columns}f', data)

                # Write the data as comma-separated values
                csv_file.write(','.join(map(str, values)) + '\n')

# Example usage
binary_file_path = 'input_file.bin'  # Replace with your binary file path
csv_file_path = 'output_file.csv'    # Replace with your desired CSV file path

convert_binary_to_csv(binary_file_path, csv_file_path)

File: bin2csv.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>

int main() {
    std::string binary_file_path = "input_file.bin"; // Replace with your binary file path
    std::string csv_file_path = "output_file.csv";   // Replace with your desired CSV file path

    std::ifstream binary_file(binary_file_path, std::ios::binary);
    std::ofstream csv_file(csv_file_path);

    if (!binary_file.is_open() || !csv_file.is_open()) {
        std::cerr << "Error opening files!" << std::endl;
        return 1;
    }

    std::string line;
    int num_columns;

    // Read the number of columns
    std::getline(binary_file, line);
    std::istringstream(line) >> num_columns;

    // Read and write the column headers
    std::getline(binary_file, line);
    csv_file << line << "\n";

    // Process the binary data
    while (!binary_file.eof()) {
        for (int i = 0; i < num_columns; ++i) {
            float value;
            binary_file.read(reinterpret_cast<char*>(&value), sizeof(value));

            if (binary_file.eof()) break;

            csv_file << value;
            if (i < num_columns - 1) csv_file << ",";
        }
        csv_file << "\n";
    }

    binary_file.close();
    csv_file.close();

    return 0;
}
